modules/core/04-channel/v2/keeper/msg_server_test.go

```diff
@@ -0,0 +1,112 @@
+	var path *ibctesting.Path
+	var msg *channeltypesv2.MsgSendPacket
+	var expectedPacket channeltypesv2.Packet
+
+	testCases := []struct {
+		name        string
+		malleate    func()
+		expError    error
+		shouldPanic bool
+	}{
+		{
+			name:     "success",
+			malleate: func() {},
+			expError: nil,
+		},
+		{
+			name: "failure: timeout elapsed",
+			malleate: func() {
+				// ensure a message timeout.
+				msg.TimeoutTimestamp = uint64(1)
+			},
+			expError: channeltypesv1.ErrTimeoutElapsed,
+		},
+		{
+			name: "failure: inactive client",
+			malleate: func() {
+				path.EndpointA.FreezeClient()
+			},
+			expError: clienttypes.ErrClientNotActive,
+		},
+		{
+			name: "failure: application callback error",
+			malleate: func() {
+				path.EndpointA.Chain.GetSimApp().MockModuleV2A.IBCApp.OnSendPacket = func(ctx context.Context, sourceID string, destinationID string, sequence uint64, data channeltypesv2.PacketData, signer sdk.AccAddress) error {
+					return mock.MockApplicationCallbackError
+				}
+			},
+			expError: mock.MockApplicationCallbackError,
+		},
+		{
+			name: "failure: counterparty not found",
+			malleate: func() {
+				msg.SourceId = "foo"
+			},
+			expError: fmt.Errorf("counterparty not found"),
+		},
+		{
+			name: "failure: route to non existing app",
+			malleate: func() {
+				msg.PacketData[0].SourcePort = "foo"
+			},
+			expError: fmt.Errorf("no route for foo"),
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		suite.Run(tc.name, func() {
+			suite.SetupTest() // reset
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			path.SetupV2()
+
+			timeoutTimestamp := suite.chainA.GetTimeoutTimestamp()
+			msg = channeltypesv2.NewMsgSendPacket(path.EndpointA.ClientID, timeoutTimestamp, suite.chainA.SenderAccount.GetAddress().String(), mockv2.NewMockPacketData(mockv2.ModuleNameA, mockv2.ModuleNameB))
+
+			expectedPacket = channeltypesv2.NewPacket(1, path.EndpointA.ClientID, path.EndpointB.ClientID, timeoutTimestamp, mockv2.NewMockPacketData(mockv2.ModuleNameA, mockv2.ModuleNameB))
+
+			tc.malleate()
+
+			res, err := path.EndpointA.Chain.SendMsgs(msg)
+
+			expPass := tc.expError == nil
+			if expPass {
+				suite.Require().NoError(err)
+				suite.Require().NotNil(res)
+
+				ck := path.EndpointA.Chain.GetSimApp().IBCKeeper.ChannelKeeperV2
+
+				packetCommitment, ok := ck.GetPacketCommitment(path.EndpointA.Chain.GetContext(), path.EndpointA.ClientID, 1)
+				suite.Require().True(ok)
+				suite.Require().Equal(channeltypesv2.CommitPacket(expectedPacket), []byte(packetCommitment), "packet commitment is not stored correctly")
+
+				nextSequenceSend, ok := ck.GetNextSequenceSend(path.EndpointA.Chain.GetContext(), path.EndpointA.ClientID)
+				suite.Require().True(ok)
+				suite.Require().Equal(uint64(2), nextSequenceSend, "next sequence send was not incremented correctly")
+
+			} else {
+				suite.Require().Error(err)
+				suite.Require().Contains(err.Error(), tc.expError.Error())
```

Comment:
We have `RequireErrorIsOrContains` in `testing/utils.go` which I think would be nicer to use here. And I think you don't need `suite.Require().Error(err)` above

Reply:
Oh yes! I forgot about that